File: python/relicpy/seek.py
"""Finding ONE session on disk, by id, without the index.

This is what makes `relic session <id>` answer for a session that was never imported:
seek it, import it, then answer. The alternative — "run index first" — is a worse
answer to a question the tool can resolve itself.
"""

from __future__ import annotations

import json
import os
from typing import Optional

from .discover import Found, _dirs, _files, _stat
from .sources import bank_of, load_sources

_cwd_cache: dict[str, Optional[str]] = {}


def cwd_of_file(file_path: str) -> Optional[str]:
    """The cwd a transcript records for itself, read from its head.

    Never decoded from the directory name: the encoding maps BOTH "/" and "." to "-",
    so it is lossy and not reversible. Reads a bounded prefix because a transcript can
    be tens of megabytes and the cwd is on an early line.
    """
    if file_path in _cwd_cache:
        return _cwd_cache[file_path]
    found = None
    try:
        with open(file_path, "r", encoding="utf-8", errors="replace") as fh:
            head = fh.read(262_144)
        for line in head.split("\n"):
            if not line.strip():
                continue
            try:
                rec = json.loads(line)
            except Exception:
                continue          # a truncated final line is expected when slicing
            cwd = rec.get("cwd") or (rec.get("payload") or {}).get("cwd") \
                if isinstance(rec, dict) else None
            if isinstance(cwd, str) and cwd:
                found = cwd
                break
    except OSError:
        pass
    _cwd_cache[file_path] = found
    return found


def seek_on_disk(session_id: str) -> list[Found]:
    """Every transcript file whose name starts with `session_id`, across every
    TRANSCRIPT source.

    Gated to transcript layouts: `claude-memory` points at the SAME directory as
    `claude-live`, so without the gate every session id matches a second time, is
    parsed by the memory shape, and lands as a bogus one-event row in the memory bank.
    """
    out: list[Found] = []
    for src in load_sources():
        if not os.path.exists(src.path):
            continue
        if src.walk not in ("claude-tiers", "flat", "omp"):
            continue
        # The same bank the bulk walker would have stamped. Without it an on-demand
        # import writes into the fallback bank instead of the source's own.
        bank = bank_of(src)

        if src.walk == "claude-tiers":
            for project in _dirs(src.path):
                pdir = os.path.join(src.path, project)
                for f in _files(pdir):
                    if not f.startswith(session_id):
                        continue
                    st = _stat(os.path.join(pdir, f))
                    if st:
                        out.append(Found(os.path.join(pdir, f), project, "session", src.key,
                                         bank, None, None, st[0], st[1], src.parser))
                for sdir in _dirs(pdir):
                    if not sdir.startswith(session_id):
                        continue
                    subagents = os.path.join(pdir, sdir, "subagents")
                    for f in _files(subagents):
                        st = _stat(os.path.join(subagents, f))
                        if st:
                            out.append(Found(os.path.join(subagents, f), project, "subagent",
                                             src.key, bank, None, f[:-6], st[0], st[1], src.parser))
                    workflows = os.path.join(subagents, "workflows")
                    for run in _dirs(workflows):
                        if not run.startswith("wf_"):
                            continue
                        for f in _files(os.path.join(workflows, run)):
                            if f == "journal.jsonl":
                                continue
                            p = os.path.join(workflows, run, f)
                            st = _stat(p)
                            if st:
                                out.append(Found(p, project, "workflow_agent", src.key, bank,
                                                 run, f[:-6], st[0], st[1], src.parser))
        else:
            stack = [src.path]
            depth = 0
            while stack and depth < 6:
                nxt = []
                for d in stack:
                    for f in _files(d):
                        if session_id not in f:
                            continue
                        st = _stat(os.path.join(d, f))
                        if st:
                            out.append(Found(os.path.join(d, f), src.key, "session", src.key,
                                             bank, None, None, st[0], st[1], src.parser))
                    nxt.extend(os.path.join(d, x) for x in _dirs(d))
                stack = nxt
                depth += 1
    return out

File: src/seek.ts
import { readdirSync, statSync, existsSync } from "node:fs";
import { join, basename } from "node:path";
import { bankOf, loadSources, transcriptRoots } from "./sources.js";
import type { Found, Tier } from "./discover.js";

/**
 * Find a transcript on DISK by session id, without consulting the index.
 *
 * This exists so `relic session <id>` never answers "run index first". A session id
 * maps to a filename, so locating it is a deterministic lookup — not something that
 * should require a human (or a model) to improvise a `find` invocation. The index is
 * an accelerator; the filesystem is the source of truth, and it is always available.
 *
 * Bounded on purpose: this matches a FILENAME PREFIX in the places transcripts are
 * known to live. It never walks the whole tree and never greps content — a full sweep
 * is what makes this class of lookup feel expensive enough to skip.
 */

const SUB = "subagents";
const WF = "workflows";

function dirs(p: string): string[] {
  try { return readdirSync(p, { withFileTypes: true }).filter(e => e.isDirectory()).map(e => e.name); }
  catch { return []; }
}
function files(p: string): string[] {
  try { return readdirSync(p, { withFileTypes: true }).filter(e => e.isFile() && e.name.endsWith(".jsonl")).map(e => e.name); }
  catch { return []; }
}
function statOf(p: string) {
  try { const s = statSync(p); return { mtime: Math.floor(s.mtimeMs / 1000), size: s.size }; }
  catch { return null; }
}

/** Every transcript whose filename starts with `id`, across all configured sources. */
export function seekOnDisk(id: string): Found[] {
  const out: Found[] = [];
  const want = (name: string) => name.startsWith(id);

  for (const src of loadSources()) {
    if (!existsSync(src.path)) continue;
    // TRANSCRIPT layouts only. `claude-memory` points at the SAME directory as
    // `claude-live`, so without this gate every session id matches a second time, is
    // parsed by parseMemory, and lands as a bogus one-event row in the memory bank.
    if (src.walk !== "claude-tiers" && src.walk !== "claude-home" && src.walk !== "flat" && src.walk !== "omp") continue;
    // Same bank the bulk walker would have stamped. Without it an on-demand
    // `relic session <id>` import writes into the fallback bank instead of the
    // source's own — a misfile that no error would report.
    const bank = bankOf(src);

    if (src.walk === "flat") {
      // Codex nests by date; bounded depth rather than an open-ended walk.
      const walk = (root: string, depth: number) => {
        for (const f of files(root)) {
          // rollout-<ts>-<uuid>.jsonl — the id is in the name, not at its start
          if (!want(f) && !f.includes(id)) continue;
          const st = statOf(join(root, f));
          if (st) out.push({ path: join(root, f), projectDir: src.key, tier: "session" as Tier,
            source: src.key, bank, workflowRunId: null, agentId: null, ...st, parser: src.parser });
        }
        if (depth >= 4) return;
        for (const d of dirs(root)) walk(join(root, d), depth + 1);
      };
      walk(src.path, 0);
      continue;
    }

    // Claude layout: check all three tiers, since a session id names a TREE.
    const projects = transcriptRoots(src).flatMap(root => dirs(root).map(name => ({ root, name })));
    for (const { root, name: project } of projects) {
      const pp = join(root, project);

      for (const f of files(pp)) {
        if (!want(f)) continue;
        const st = statOf(join(pp, f));
        if (st) out.push({ path: join(pp, f), projectDir: project, tier: "session",
          source: src.key, bank, workflowRunId: null, agentId: null, ...st, parser: src.parser });
      }

      // children live under <uuid>/subagents/... — so the DIRECTORY carries the id
      for (const sessionDir of dirs(pp)) {
        if (!want(sessionDir)) continue;
        const sub = join(pp, sessionDir, SUB);
        if (!existsSync(sub)) continue;

        for (const f of files(sub)) {
          const st = statOf(join(sub, f));
          if (st) out.push({ path: join(sub, f), projectDir: project, tier: "subagent",
            source: src.key, bank, workflowRunId: null, agentId: basename(f, ".jsonl"), ...st, parser: src.parser });
        }
        const wf = join(sub, WF);
        for (const run of dirs(wf)) {
          if (!run.startsWith("wf_")) continue;
          for (const f of files(join(wf, run))) {
            // journal.jsonl is the workflow RUNNER's event log, not a transcript.
            // discover.ts skips it; this path did not, so an on-demand
            // `relic session <id>` imported one bogus workflow_agent row per run.
            // Found by porting seek to Python: 162 files there against 169 here,
            // and this session has exactly 7 wf_ runs.
            if (f === "journal.jsonl") continue;
            const st = statOf(join(wf, run, f));
            if (st) out.push({ path: join(wf, run, f), projectDir: project, tier: "workflow_agent",
              source: src.key, bank, workflowRunId: run, agentId: basename(f, ".jsonl"), ...st, parser: src.parser });
          }
        }
      }
    }
  }
  return out;
}
